GSA-TTS/FAC
# Search - Pagination

# Search - Pagination

Issue, best fit: https://github.com/GSA-TTS/FAC/issues/2236 or https://github.com/GSA-TTS/FAC/issues/2600

## Changes:
Pagination:
1. A `limit` (default 30) and `page` value in the form. Defines how many results show on the page, and what the offset of results are.
2. Results are passed into a pagination object. This object then:
    * Splices the results for the _nth_ page `[(n-1) * lim : (n-1) * lim + lim]`. For example: page 3, limit 30 gives results 61-90.
    * Defines the pagination sub elements (1, 2, ..., 7, 8, 9, ..., 19, 20), including ellipsis.
3. HTML displays the pagination sub elements nicely
4. JS overrides the pagination links:
    * Clicking "Next page" or "Previous page" add/subtracts one from the form page element, the submits the form for a reload
    * Clicking any number sets the form page element to that number, then submits the form for a reload

Small things:
1. Center the PDF view icon
3. Match the Audit Year checkbox background color to the filters color (color-base-lighter, #efefef)
6. By default, sort results by FAC accepted date. It needs some sort of strict sort field, or pages will be wonky between each other

## How to test:
Switch to this branch and run normally. Make sure a good amount of items are available in the dissemination tables. I do this by running something like:
```
python manage.py end_to_end_test_data_generator --email <some_exisiting_email_here> --dbkeys "101078,101101,101352,101382,101546,101579,101590,101590,101893,10213,102138,102184" --years "22,22,22,22,22,22,22,22,22,22,22,22"
```
This puts twelve submissions all the way through to dissemination. It takes a little while.

To avoid having to put literally hundreds of submissions through e2e for testing, you can & should edit the default page limit of 30 to something smaller, like two, for demonstration. Do this at `backend/dissemination/views.py`, line ~48. 

Now, check out search at http://localhost:8000/dissemination/search/. Run any search that gets results, like the blank one. Verify that pages:
1. Exist. 
2. Are navigable by the "Next" and "Previous" buttons.
3. Are navigable by the page numbers.
4. Display correct information (different results on each page, no copies).
5. Display the correct pagination links at the bottom (ellipses when there are lots of pages).

## Screenshots:
One total page:
![image](https://github.com/GSA-TTS/FAC/assets/91098850/8edc2904-6951-43ac-ba69-86f71995c201)

Two pages, on the first page:
![image](https://github.com/GSA-TTS/FAC/assets/91098850/1c20ba17-8298-40aa-aa5d-7be210daae57)

Two pages, on the second page:
![image](https://github.com/GSA-TTS/FAC/assets/91098850/522ccfb2-687e-416a-b2f1-3a29c79b2019)

Many pages, on the first one:
![image](https://github.com/GSA-TTS/FAC/assets/91098850/8dd5fb94-c153-49f5-a17a-7efc2cc9b0b4)

Many pages, in the middle:
![image](https://github.com/GSA-TTS/FAC/assets/91098850/4785ff4a-bdc7-4958-b7f2-643374eddccc)

Many pages, at the end:
![image](https://github.com/GSA-TTS/FAC/assets/91098850/8d5ba132-2413-4696-bf15-e96e3a6d4e04)


## PR checklist: submitters


## PR checklist: reviewers


The larger the PR, the stricter we should be about these points.
